File: YoiSceneKirinuKu/Preflight.swift
import Foundation
import CoreFoundation

enum PreflightErrorCode: String, Codable, CaseIterable, Sendable {
    case invalidRequest = "invalid_request"
    case unsupportedFileType = "unsupported_file_type"
    case inputNotFound = "input_not_found"
    case inputNotReadable = "input_not_readable"
    case probeNotStarted = "probe_not_started"
    case probeTimedOut = "probe_timed_out"
    case probeFailed = "probe_failed"
    case invalidProbeOutput = "invalid_probe_output"
    case videoStreamMissing = "video_stream_missing"
    case audioStreamMissing = "audio_stream_missing"
    case invalidDuration = "invalid_duration"
    case protocolError = "protocol_error"
    case internalError = "internal_error"

    var userMessage: String {
        switch self {
        case .invalidRequest, .probeNotStarted, .internalError:
            "動画を確認できませんでした"
        case .unsupportedFileType:
            "MP4ファイルを選んでください"
        case .inputNotFound:
            "選択した動画が見つかりません"
        case .inputNotReadable:
            "この動画を読み込めません"
        case .probeTimedOut:
            "動画の確認に時間がかかりすぎました"
        case .probeFailed:
            "この動画は解析できません"
        case .invalidProbeOutput:
            "動画情報を確認できませんでした"
        case .videoStreamMissing:
            "映像トラックが見つかりません"
        case .audioStreamMissing:
            "音声トラックが見つかりません"
        case .invalidDuration:
            "動画の長さを確認できません"
        case .protocolError:
            "動画の確認処理で問題が発生しました"
        }
    }
}

struct PreflightResult: Equatable, Sendable {
    let fileName: String
    let durationMilliseconds: Int64
    let containerFormat: String
    let videoStreamCount: Int
    let audioStreamCount: Int
}

enum PreflightOutcome: Equatable, Sendable {
    case success(PreflightResult)
    case failure(PreflightErrorCode)
}

protocol PreflightServicing: Sendable {
    func run(sourceURL: URL, requestID: UUID) async -> PreflightOutcome
}

struct PreflightConfiguration: Sendable {
    let pythonExecutableURL: URL
    let ffprobeExecutableURL: URL
    let scriptURL: URL

    static func bundled(bundle: Bundle = .main) -> PreflightConfiguration? {
        guard
            let pythonPath = bundle.object(forInfoDictionaryKey: "PreflightPythonExecutable") as? String,
            let ffprobePath = bundle.object(forInfoDictionaryKey: "PreflightFFprobeExecutable") as? String,
            let scriptURL = bundle.url(forResource: "preflight", withExtension: "py")
        else {
            return nil
        }

        return PreflightConfiguration(
            pythonExecutableURL: URL(fileURLWithPath: pythonPath),
            ffprobeExecutableURL: URL(fileURLWithPath: ffprobePath),
            scriptURL: scriptURL
        )
    }
}

final class PreflightService: PreflightServicing, @unchecked Sendable {
    private let configuration: PreflightConfiguration?

    init(configuration: PreflightConfiguration? = .bundled()) {
        self.configuration = configuration
    }

    func run(sourceURL: URL, requestID: UUID) async -> PreflightOutcome {
        guard let configuration else {
            return .failure(.probeNotStarted)
        }

        let controller = PreflightProcessController()
        return await withTaskCancellationHandler {
            await withCheckedContinuation { continuation in
                DispatchQueue.global(qos: .userInitiated).async {
                    continuation.resume(returning: Self.execute(
                        sourceURL: sourceURL,
                        requestID: requestID,
                        configuration: configuration,
                        controller: controller
                    ))
                }
            }
        } onCancel: {
            controller.terminate()
        }
    }

    private static func execute(
        sourceURL: URL,
        requestID: UUID,
        configuration: PreflightConfiguration,
        controller: PreflightProcessController
    ) -> PreflightOutcome {
        guard validateExecutable(configuration.pythonExecutableURL),
              validateExecutable(configuration.ffprobeExecutableURL),
              FileManager.default.fileExists(atPath: configuration.scriptURL.path) else {
            return .failure(.probeNotStarted)
        }

        let process = Process()
        let inputPipe = Pipe()
        let outputPipe = Pipe()
        let errorPipe = Pipe()
        process.executableURL = configuration.pythonExecutableURL
        process.arguments = [configuration.scriptURL.path, configuration.ffprobeExecutableURL.path]
        process.standardInput = inputPipe
        process.standardOutput = outputPipe
        process.standardError = errorPipe
        guard controller.register(process) else {
            return .failure(.internalError)
        }

        let request: [String: Any] = [
            "protocol_version": 1,
            "request_id": requestID.uuidString.lowercased(),
            "operation": "preflight",
            "source_path": sourceURL.path,
        ]

        guard let requestData = try? JSONSerialization.data(withJSONObject: request) else {
            return .failure(.internalError)
        }

        do {
            try process.run()
        } catch {
            controller.clear(process)
            return .failure(.probeNotStarted)
        }
        if controller.isCancellationRequested {
            controller.terminate()
        }

        let streamGroup = DispatchGroup()
        let dataStore = PreflightStreamData()
        streamGroup.enter()
        DispatchQueue.global(qos: .userInitiated).async {
            dataStore.setStdout(outputPipe.fileHandleForReading.readDataToEndOfFile())
            streamGroup.leave()
        }
        streamGroup.enter()
        DispatchQueue.global(qos: .utility).async {
            dataStore.setStderr(errorPipe.fileHandleForReading.readDataToEndOfFile())
            streamGroup.leave()
        }

        do {
            try inputPipe.fileHandleForWriting.write(contentsOf: requestData)
            try inputPipe.fileHandleForWriting.write(contentsOf: Data([0x0A]))
            try inputPipe.fileHandleForWriting.close()
        } catch {
            controller.terminate()
        }

        let deadline = DispatchTime.now() + .seconds(30)
        while process.isRunning && DispatchTime.now() < deadline {
            Thread.sleep(forTimeInterval: 0.02)
        }
        if process.isRunning {
            controller.terminate()
        }
        process.waitUntilExit()
        streamGroup.wait()
        controller.clear(process)

        guard !Task.isCancelled else {
            return .failure(.internalError)
        }

        return PreflightProtocolParser.parse(
            stdout: dataStore.stdout,
            requestID: requestID,
            terminationStatus: process.terminationStatus,
            terminationReason: process.terminationReason
        )
    }

    private static func validateExecutable(_ url: URL) -> Bool {
        url.path.hasPrefix("/")
            && FileManager.default.fileExists(atPath: url.path)
            && FileManager.default.isExecutableFile(atPath: url.path)
    }
}

enum PreflightProtocolParser {
    static func parse(
        stdout: Data,
        requestID: UUID,
        terminationStatus: Int32,
        terminationReason: Process.TerminationReason
    ) -> PreflightOutcome {
        guard terminationReason == .exit, terminationStatus == 0,
              let text = String(data: stdout, encoding: .utf8) else {
            return .failure(.protocolError)
        }

        let lines = text.split(separator: "\n", omittingEmptySubsequences: false)
        guard lines.last == "" else {
            return .failure(.protocolError)
        }

        var expectedSequence = 1
        var errorCode: PreflightErrorCode?
        var errorRawCode: String?
        var terminalOutcome: PreflightOutcome?
        var didFinish = false

        for line in lines.dropLast() {
            guard !line.isEmpty,
                  !didFinish,
                  let data = line.data(using: .utf8),
                  let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  object["protocol_version"] as? Int == 1,
                  object["request_id"] as? String == requestID.uuidString.lowercased(),
                  object["sequence"] as? Int == expectedSequence,
                  let type = object["type"] as? String,
                  let payload = object["payload"] as? [String: Any]
            else {
                return .failure(.protocolError)
            }
            expectedSequence += 1

            switch type {
            case "progress":
                guard payload["stage"] as? String == "preflight",
                      ["running", "completed"].contains(payload["status"] as? String) else {
                    return .failure(.protocolError)
                }
            case "error":
                guard errorCode == nil,
                      let rawCode = payload["code"] as? String else {
                    return .failure(.protocolError)
                }
                errorRawCode = rawCode
                errorCode = PreflightErrorCode(rawValue: rawCode) ?? .internalError
            case "finished":
                didFinish = true
                guard let outcome = payload["outcome"] as? String else {
                    return .failure(.protocolError)
                }
                if outcome == "failed" {
                    guard let rawCode = payload["code"] as? String,
                          rawCode == errorRawCode,
                          let code = errorCode else {
                        return .failure(.protocolError)
                    }
                    terminalOutcome = .failure(code)
                } else if outcome == "succeeded" {
                    guard errorCode == nil,
                          let result = parseResult(payload["result"]) else {
                        return .failure(.protocolError)
                    }
                    terminalOutcome = .success(result)
                } else {
                    return .failure(.protocolError)
                }
            default:
                return .failure(.protocolError)
            }
        }

        return terminalOutcome ?? .failure(.protocolError)
    }

    private static func parseResult(_ value: Any?) -> PreflightResult? {
        guard let value = value as? [String: Any],
              let fileName = value["file_name"] as? String, !fileName.isEmpty,
              let duration = exactPositiveInteger(value["duration_ms"]),
              let format = value["container_format"] as? String, !format.isEmpty,
              let videoCountValue = exactPositiveInteger(value["video_stream_count"]),
              let audioCountValue = exactPositiveInteger(value["audio_stream_count"]),
              videoCountValue <= Int64(Int.max),
              audioCountValue <= Int64(Int.max) else {
            return nil
        }
        return PreflightResult(
            fileName: fileName,
            durationMilliseconds: duration,
            containerFormat: format,
            videoStreamCount: Int(videoCountValue),
            audioStreamCount: Int(audioCountValue)
        )
    }

    private static func exactPositiveInteger(_ value: Any?) -> Int64? {
        guard let number = value as? NSNumber,
              CFGetTypeID(number) != CFBooleanGetTypeID(),
              number.doubleValue.isFinite,
              number.doubleValue == Double(number.int64Value),
              number.int64Value > 0 else {
            return nil
        }
        return number.int64Value
    }
}

private final class PreflightProcessController: @unchecked Sendable {
    private let lock = NSLock()
    private var process: Process?
    private var cancellationRequested = false

    var isCancellationRequested: Bool {
        lock.withLock { cancellationRequested }
    }

    func register(_ process: Process) -> Bool {
        lock.withLock {
            guard !cancellationRequested else { return false }
            self.process = process
            return true
        }
    }

    func clear(_ process: Process) {
        lock.withLock {
            if self.process === process { self.process = nil }
        }
    }

    func terminate() {
        let registeredProcess = lock.withLock {
            cancellationRequested = true
            return process
        }
        if registeredProcess?.isRunning == true { registeredProcess?.terminate() }
    }
}

private final class PreflightStreamData: @unchecked Sendable {
    private let lock = NSLock()
    private var storedStdout = Data()
    private var storedStderr = Data()

    var stdout: Data { lock.withLock { storedStdout } }
    var stderr: Data { lock.withLock { storedStderr } }

    func setStdout(_ data: Data) { lock.withLock { storedStdout = data } }
    func setStderr(_ data: Data) { lock.withLock { storedStderr = data } }
}

File: YoiSceneKirinuKu/preflight.py
#!/usr/bin/env python3
import json
import math
import os
from pathlib import Path
import subprocess
import sys

PROTOCOL_VERSION = 1
ERROR_CODES = {
    "invalid_request",
    "unsupported_file_type",
    "input_not_found",
    "input_not_readable",
    "probe_not_started",
    "probe_timed_out",
    "probe_failed",
    "invalid_probe_output",
    "video_stream_missing",
    "audio_stream_missing",
    "invalid_duration",
    "internal_error",
}


class EventWriter:
    def __init__(self, request_id):
        self.request_id = request_id
        self.sequence = 0

    def write(self, event_type, payload):
        self.sequence += 1
        event = {
            "protocol_version": PROTOCOL_VERSION,
            "type": event_type,
            "request_id": self.request_id,
            "sequence": self.sequence,
            "payload": payload,
        }
        sys.stdout.write(json.dumps(event, ensure_ascii=False, separators=(",", ":")) + "\n")
        sys.stdout.flush()

    def fail(self, code):
        if code not in ERROR_CODES:
            code = "internal_error"
        self.write("error", {"code": code})
        self.write("finished", {"outcome": "failed", "code": code})


def valid_request(value):
    return (
        isinstance(value, dict)
        and set(value) == {"protocol_version", "request_id", "operation", "source_path"}
        and value.get("protocol_version") == PROTOCOL_VERSION
        and value.get("operation") == "preflight"
        and isinstance(value.get("request_id"), str)
        and isinstance(value.get("source_path"), str)
        and os.path.isabs(value["source_path"])
    )


def main():
    raw = sys.stdin.buffer.readline()
    try:
        request = json.loads(raw)
    except (json.JSONDecodeError, UnicodeDecodeError):
        return 2

    if not valid_request(request) or sys.stdin.buffer.read(1):
        return 2

    writer = EventWriter(request["request_id"])
    source = Path(request["source_path"])
    if source.suffix.lower() != ".mp4":
        writer.fail("unsupported_file_type")
        return 0
    if not source.exists() or not source.is_file():
        writer.fail("input_not_found")
        return 0
    if not os.access(source, os.R_OK):
        writer.fail("input_not_readable")
        return 0

    ffprobe = Path(sys.argv[1]) if len(sys.argv) == 2 else None
    if ffprobe is None or not ffprobe.is_absolute() or not ffprobe.is_file() or not os.access(ffprobe, os.X_OK):
        writer.fail("probe_not_started")
        return 0

    writer.write("progress", {"stage": "preflight", "status": "running"})
    arguments = [
        str(ffprobe), "-v", "error", "-print_format", "json",
        "-show_format", "-show_streams", "--", str(source),
    ]
    try:
        completed = subprocess.run(arguments, shell=False, capture_output=True, timeout=20, check=False)
    except subprocess.TimeoutExpired:
        writer.fail("probe_timed_out")
        return 0
    except OSError:
        writer.fail("probe_not_started")
        return 0

    if completed.returncode != 0:
        writer.fail("probe_failed")
        return 0
    try:
        probe = json.loads(completed.stdout)
    except (json.JSONDecodeError, UnicodeDecodeError):
        writer.fail("invalid_probe_output")
        return 0

    streams = probe.get("streams")
    media_format = probe.get("format")
    if not isinstance(streams, list) or not isinstance(media_format, dict):
        writer.fail("invalid_probe_output")
        return 0
    format_name = media_format.get("format_name")
    if not isinstance(format_name, str) or "mp4" not in format_name.split(","):
        writer.fail("invalid_probe_output")
        return 0
    video_count = sum(stream.get("codec_type") == "video" for stream in streams if isinstance(stream, dict))
    audio_count = sum(stream.get("codec_type") == "audio" for stream in streams if isinstance(stream, dict))
    if video_count == 0:
        writer.fail("video_stream_missing")
        return 0
    if audio_count == 0:
        writer.fail("audio_stream_missing")
        return 0
    try:
        duration_seconds = float(media_format["duration"])
    except (KeyError, TypeError, ValueError):
        writer.fail("invalid_duration")
        return 0
    if not math.isfinite(duration_seconds) or duration_seconds <= 0:
        writer.fail("invalid_duration")
        return 0

    duration_ms = math.ceil(duration_seconds * 1000)
    writer.write("progress", {"stage": "preflight", "status": "completed"})
    writer.write("finished", {
        "outcome": "succeeded",
        "result": {
            "file_name": source.name,
            "duration_ms": duration_ms,
            "container_format": format_name,
            "video_stream_count": video_count,
            "audio_stream_count": audio_count,
        },
    })
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
